### PR TITLE
analytic filtering mechanism in mis_builder (was mis_builder_analytic_filter)

### LINKED ISSUE #15: analytic filters
See previous work
- https://github.com/OCA/account-financial-reporting/pull/243 now in #24 
- https://github.com/OCA/account-financial-reporting/pull/233
- https://github.com/OCA/account-financial-reporting/pull/168 now in #29

### LINKED ISSUE #42: Control visibility of analytic filter combo box in widget
On mis.report.instance, add a flag to control the visibility of the analytic filter combo box in the widget. 

For some report (eg Balance Sheet), the analytic filter makes no sense, while on other reports it does. So it must be controllable on a per instance basis, and not only with the security group.

### LINKED ISSUE #44: Show analytic filter in exported pdf and xls
This could be done with a `filter_description` computed field on `mis.report.instance`.

This field could then be displayed:
* in the Filters tab on the mis.report.instance form
* in the Excel export
* in the PDF export

### PR DESCRIPTION
**Pull request by adrienpeiffer**
Include https://github.com/OCA/account-financial-reporting/pull/243 into mis_builder module

closes #15
closes #42 
closes #44

### CODE REVIEW

File: mis_builder/models/mis_report_instance.py [hunk 1]
```diff
@@ -467,6 +476,7 @@ def preview(self):
             'view_type': 'form',
             'view_id': view_id.id,
             'target': 'current',
+            'context': context
```
Comment (lepistone) [inline mis_builder/models/mis_report_instance.py]: I suggest a comma here

**Review by lepistone — APPROVED** [review]

**Comment by adrienpeiffer** [comment]
@lepistone done :smile:

File: mis_builder/models/mis_report_instance.py [hunk 2]
```diff
@@ -376,6 +380,8 @@ def _default_company(self):
     date_from = fields.Date(string="From")
     date_to = fields.Date(string="To")
     temporary = fields.Boolean(default=False)
+    account_analytic_id = fields.Many2one(
+        comodel_name='account.analytic.account', string='Analytic Account')
```
Comment (sbidoul) [inline mis_builder/models/mis_report_instance.py]: @adrienpeiffer there is a mix of `analytic_account_id` and `account_analytic_id`.

Could we name the field `analytic_account_id` as in `mis_builder_budget` with an oldname attribute?

**Comment by adrienpeiffer** [comment]
@sbidoul Done and rebased

File: mis_builder/views/mis_report_instance.xml [hunk 3]
```diff
@@ -81,6 +81,7 @@
                                 <field name="date_range_id"/>
                                 <field name="date_from" attrs="{'required': [('comparison_mode', '=', False)]}"/>
                                 <field name="date_to" attrs="{'required': [('comparison_mode', '=', False)]}"/>
+                                <field name="analytic_account_id" groups="analytic.group_analytic_accounting"/>
```
Comment (sbidoul) [inline mis_builder/views/mis_report_instance.xml]: This field must move out of this group, because it's hidden in comparison mode.

Comment (adrienpeiffer) [reply mis_builder/views/mis_report_instance.xml]: Well spotted !

**Comment by sbidoul** [comment]
We'll need a test for this.

File: mis_builder/static/src/xml/mis_widget.xml [hunk 4]
```diff
@@ -9,6 +9,7 @@
                     <button class="oe_mis_builder_export btn btn-sm oe_button"><span class="fa fa-download"/>Export</button>
                     <button style="display: none;" class="oe_mis_builder_settings btn btn-sm oe_button"><span class="fa fa-cog"/> Settings</button>
                 </div>
+                <div class="oe_mis_builder_analytic_account" style="visibility: hidden; position: relative; display: inline-block;"></div>
```
Comment (sbidoul) [inline mis_builder/static/src/xml/mis_widget.xml]: should we call this oe_mis_builder_filterbox or something?

**Review by benwillig — APPROVED** [review]
Code review and test :+1:

**Comment by sbidoul** [comment]
I rebased and moved the filter field to the new filters group in the report instance form.

**Comment by jalzaga** [comment]
When you print in pdf it does not apply the analytic account filter, when you export to excel it is correctly applied.

**Comment by jalzaga** [comment]
If you have a query in your report, how do you filter by the selected analytic account? I guess you have to put it in the domain, how do you reference it?

**Comment by sbidoul** [comment]
@jalzaga thanks for testing!

To have analytic filter work on queries, you need to develop an addon module that overrides _get_additional_query_filter: https://github.com/OCA/mis-builder/blob/edc61e986664f74d50dfaa71ed23e17fadddeb19/mis_builder/models/mis_report_instance.py#L305-L316

**Comment by jalzaga** [comment]
Thanks for your prompt response @sbidoul 
It would be a great improvement to make the analytic filter many2many to select more than one analytic account, I think it would not be too hard to implement. 
Appart from that, is it in your plans to add analytic tags?.

**Comment by jalzaga** [comment]
When you print in pdf or export to excel the analytic account filter does not appear in them. It would be very useful.

**Comment by jalzaga** [comment]
@sbidoul about te query filter, this would be the method, I have tested it and it works:
    @api.multi
    def _get_additional_query_filter(self, query):
        """ Prepare an additional filter to apply on the query

        This filter is combined to the query domain with a AND
        operator. This hook is intended
        to be inherited, and is useful to implement filtering
        on analytic dimensions or operational units.

        Returns an Odoo domain expression (a python list)
        compatible with the model of the query."""
        self.ensure_one()
        filters = []
        analytic_account_id = self.env.context.get('analytic_account_id')
        if analytic_account_id:
            filters.append(('analytic_account_id', '=', analytic_account_id))
        return filters

**Comment by rafaelbn** [comment]
Hi @adrienpeiffer @sbidoul @jbeficent ,

What do you think about merging this PR? :+1:

**Comment by sbidoul** [comment]
There are still a few improvements I would like to do before merging. No functional change but a more stable api for people extending it.

- [ ] a test
- [ ] renaming the div as mis_builder_filterbox in the widget so it's more consistent when adding additional filters
- [ ] checking the JS api is compatible with the v11 widget (which we plan to backport to 10) so we provide a stable JS api that's easy to extend

We'll probably to that with @benwillig in the coming weeks

**Comment by rafaelbn** [comment]
OK, thank you!!!

**Comment by Stephanvandenberg** [comment]
Hi all,

I'm new here, so please forgive me if this is not the right place for this comment. 

We are using the analytic filter pull request. It works great but there are some possible improvements. 
- analytic account many2many filter instead of many2one
- analytic account tag filter 

Are any of these changes planned?

Best regards,

Stephan

**Comment by sbidoul** [comment]
@Stephanvandenberg yes, it's a good place to ask. You are most welcome.

What you ask is not immediately planned, but it makes sense, although it's not trivial to integrate analytics in a way that fits all use cases. Many enhancement requests related to analytic filtering are currently grouped in [milestone 3.3](https://github.com/OCA/mis-builder/milestone/3). Next time we have some budget related to this topic we'll finish it. It will come sooner or later depending on financing.

In the meantime, I suggest you create to separate issues with your topics so we can keep track of them.

**Comment by HaraldPanten** [comment]
Hi @sbidoul , @lepistone , @rafaelbn 

How's this PR going? I've been working with it and it's working fine. Even though, should it be implemented in the mis_builder module, or should we add this functionality in a new one, like it was in V8?

What do you think?

Let me know if I can help you.

THX!

**Comment by AaronHForgeFlow** [comment]
something when wrong here, there are conflicts  :face_with_head_bandage:

**Comment by benwillig** [comment]
@aheficent It's because we backported the widget from the v11. All should be ok now. @sbidoul Could you review please?

**Review by AaronHForgeFlow — APPROVED** [review]
@benwillig All right. It was working before so from my side this is fine. Thanks!

**Comment by benwillig** [comment]
At the moment, filters doesn't work when exporting or printing the report. I'm currently working on that.

**Comment by sbidoul** [comment]
@benwillig @sbejaoui Can you also add a file named `15.feature` en `mis_builder/readme/newfragments`, with the text of the changelog entry. #15  being the corresponding issue number.

**Comment by benwillig** [comment]
Added a commit which fix https://github.com/OCA/mis-builder/issues/44

**Comment by benwillig** [comment]
Added a commit which fix https://github.com/OCA/mis-builder/issues/42

File: mis_builder/models/mis_report_instance.py [hunk 5]
```diff
@@ -300,9 +300,10 @@ def _get_additional_move_line_filter(self):
         compatible with account.move.line."""
         self.ensure_one()
         filters = []
-        analytic_account_id = self.env.context.get('analytic_account_id')
-        if analytic_account_id:
-            filters.append(('analytic_account_id', '=', analytic_account_id))
+        mis_report_filters = self.env.context.get('mis_report_filters', {})
+        for filter, value in mis_report_filters.iteritems():
```
Comment (sbidoul) [inline mis_builder/models/mis_report_instance.py]: filter is a reserved word?
s/iteritems/items/ for python 3 compat

File: mis_builder/models/mis_report_instance.py [hunk 6]
```diff
@@ -445,6 +445,9 @@ def _default_company_id(self):
     analytic_account_id = fields.Many2one(
         comodel_name='account.analytic.account', string='Analytic Account',
         oldname='account_analytic_id')
+    filter_description = fields.Char(
+        compute='_compute_filter_description'
+    )
```
Comment (sbidoul) [inline mis_builder/models/mis_report_instance.py]: Use the method from the template, no need for a field.

File: mis_builder_demo/models/mis_committed_purchase.sql [hunk 7]
```diff
@@ -42,6 +43,7 @@ CREATE OR REPLACE VIEW mis_committed_purchase AS (
         ail.name AS name,
         ail.create_date as date,
         ail.account_id as account_id,
+        NULL as analytic_account_id,
```
Comment (sbidoul) [inline mis_builder_demo/models/mis_committed_purchase.sql]: get the real field

**Comment by benwillig** [comment]
This PR is ready to be reviewed now.

**Review by HaraldPanten — APPROVED** [review]
Functional review :+1:

File: mis_builder_demo/models/mis_committed_purchase.sql [hunk 8]
```diff
@@ -54,4 +56,4 @@ CREATE OR REPLACE VIEW mis_committed_purchase AS (
             LEFT JOIN account_invoice ai ON ai.id = ail.invoice_id
         WHERE ai.state = 'draft' AND ai.type IN ('in_invoice', 'in_refund')
     ) AS mis_committed_purchase
-)
+)
```
Comment (elicoidal) [inline mis_builder_demo/models/mis_committed_purchase.sql]: empty line

File: mis_builder/static/src/js/mis_report_widget.js [hunk 9]
```diff
@@ -1,13 +1,17 @@
-odoo.define('mis_builder.widget', function(require) {
-"use strict";
+odoo.define('mis_builder.widget', function (require) {
```
Comment (elicoidal) [inline mis_builder/static/src/js/mis_report_widget.js]: add copyright

File: mis_builder/models/mis_report_instance.py [hunk 10]
```diff
@@ -437,6 +442,12 @@ def _default_company_id(self):
     date_from = fields.Date(string="From")
     date_to = fields.Date(string="To")
     temporary = fields.Boolean(default=False)
+    analytic_account_id = fields.Many2one(
+        comodel_name='account.analytic.account', string='Analytic Account',
+        oldname='account_analytic_id')
+    hide_analytic_filters = fields.Boolean(
+        string="Hide analytic filters",
```
Comment (elicoidal) [inline mis_builder/models/mis_report_instance.py]: no need string

**Review by elicoidal — COMMENTED** [review]
is there a need for version bump?

**Comment by HaraldPanten** [comment]
Hi all,

Keeping in mind  these new commits, is this PR ready to merge?

Thanks!

**Comment by benwillig** [comment]
Yes this PR is ready to be merged.

**Comment by sbidoul** [comment]
I'll do some additional review and tests in the coming days, but it should be stable now.

**Comment by sbidoul** [comment]
Rebased, squashed a bit and made some additional minor fixes and improvements.

@benwillig would you like to review/test the last changes?

Then I'll merge and back/forward port.

File: mis_builder/models/mis_report_instance.py [hunk 11]
```diff
@@ -554,16 +554,26 @@ def _onchange_dates(self):
                     self.date_to != self.date_range_id.date_end:
                 self.date_range_id = False
 
+    @api.multi
+    def _add_analytic_filters_to_context(self, context):
+        if self.analytic_account_id:
```
Comment (benwillig) [inline mis_builder/models/mis_report_instance.py:574]: ```ensure_one``` is missing here.

File: mis_builder/models/mis_report_instance.py [hunk 12]
```diff
@@ -554,16 +554,26 @@ def _onchange_dates(self):
...
+            context['mis_report_filters']['analytic_account_id'] = \
+                self.analytic_account_id.id
+
+    @api.multi
+    def _context_with_filters(self):
+        if 'mis_report_filters' in self.env.context:
```
Comment (benwillig) [inline mis_builder/models/mis_report_instance.py:582]: I believe that we need an ensure_one here too.

**Review by benwillig — CHANGES_REQUESTED** [review]
There is just missing ensure_one calls, otherwise it's ok for me.

**Comment by sbidoul** [comment]
Anyone knows why the travis job get stuck here?

I'm trying to run a full pdf report (including wkhtmltopdf) in 99eadfd77dbdeebbafa2e688b4b7e4e305c57425

These tests pass when I run them on my machine.

**Comment by sbidoul** [comment]
I suspect it blocks when wkhtmltopdf tries to http get the report assets but I don't understand why it does not work on travis.

**Comment by pedrobaeza** [comment]
Exactly, the problem comes from wkhtmltopdf locking DB in parallel with Odoo, so the only solution is to put `HttpCase` in tests so that a new cursor is put. This is another pain testing in integrated environments, needing that all possible run tests should be HttpCase for avoiding this case.

**Comment by sbidoul** [comment]
Pedro, thanks. I figured that too in the end. Just wondering why it did work on my laptop and not on Travis...

**Comment by pedrobaeza** [comment]
Maybe your install/test procedure don't follow the same steps as in Travis. We try to normalize this in https://github.com/odoo/odoo/pull/27883

**Comment by lmignon** [comment]
@sbidoul I also facing the same problem with a PR on shopinvader. Even if a launch the same commands on my laptop than those launched into travis,  I'm not able to reproduce the problem. The only way I can reproduce the problem is by running the test into a travis docker image... In my case HttpCase doesn't solve my issue.  I'm the feeling tat we miss something...

**Comment by pedrobaeza** [comment]
@lmignon please point me to the PR to see if I can tell you.

**Comment by sbidoul** [comment]
odoo/odoo#27883 is nice indeed but does not seem to be related

**Comment by lmignon** [comment]
@pedrobaeza It's outside OCA (https://github.com/akretion/odoo-shopinvader/pull/201) but uses the OCA scripts https://travis-ci.org/akretion/odoo-shopinvader/jobs/438521620#L2956

File: mis_builder/models/mis_report_instance.py [hunk 13]
```diff
@@ -533,6 +557,22 @@ def _onchange_dates(self):
                     self.date_to != self.date_range_id.date_end:
                 self.date_range_id = False
 
+    @api.multi
+    def _add_analytic_filters_to_context(self, context):
+        if self.analytic_account_id:
```
Comment (mpanarin) [inline mis_builder/models/mis_report_instance.py:574]: multi with a `self` usage can be dangerous

File: mis_builder/static/src/css/custom.css [hunk 14]
```diff
@@ -28,3 +28,11 @@
...
   padding: 10px;
 }
+
+.oe_mis_builder_analytic_filter_box {
+  padding: 10px;
+}
+
+.oe_mis_builder_buttons {
+  margin: 0 !important;
+}
```
Comment (mpanarin) [inline mis_builder/static/src/css/custom.css:38]: missing newline

**Review by mpanarin — APPROVED** [review]
2 minor comments

File: mis_builder/models/mis_report_instance.py (hunk 13 above)
Comment (sebalix) [reply mis_builder/models/mis_report_instance.py:574]: Fixed in https://github.com/acsone/mis-builder/pull/2

**Review by sebalix — APPROVED** [review]
LGTM, I fix the missing `ensure_one` calls in https://github.com/acsone/mis-builder/pull/2

**Comment by sbidoul** [comment]
Thanks all. The last thing that annoys me a tiny bit is the slight lack of test coverage in the modified parts. But I think I'll live with it.

So merging. I'll do the forward and back port in the coming days.

**Comment by HaraldPanten** [comment]
Good! :+1:

**Review by jalzaga — APPROVED** [review]
Thanks fo the PR
Changes tested. Everything OK. Only Two things to remark:
1. Analytic accounts can be created and updated from the report preview. Consider to prevent these updates since this is a reporting functionality. 
2. The filter is not applied to budget lines. Maybe this is not in the scope of this PR since there is another module for mis budget. I guess it would be in another PR for the mis builder budget module.

**Comment by sbidoul** [comment]
@jalzaga thanks for the test! It is definitely in scope for this PR to filter on budget items too. 
bcbf971 should fix this, can you test again?

@benwillig can you review bcbf971 too?

@benwillig do you know how to disable creation on the filter m2o widgets?

**Review by jalzaga — CHANGES_REQUESTED** [review]
Tested. 
The filter is correctly applied but there is a problem:  If you select an analytic account and refresh and you blank out the analytic filter and refresh again, the report doesn't show figures for entries with analytic account, it only shows the ones that have analytic account. It seems as if it selects only  moves (account moves and budgets) with no analytic account.

**Comment by benwillig** [comment]
@sbidoul Yes I know what to do to disable the create option.

@jalzaga I see what I need to change, I will fix it today I think.

File: mis_builder_budget/models/mis_report_instance_period.py [hunk 15]
```diff
@@ -32,7 +32,25 @@ def _get_additional_budget_item_filter(self):
...
         self.ensure_one()
-        return []
+        filters = []
+        mis_report_filters = self.env.context.get('mis_report_filters', {})
+        for filter_name, values in mis_report_filters.items():
+            if values:
+                value = values.get('value')
+                operator = values.get('operator', '=')
+                filters.append((filter_name, operator, value))
+        return filters
```
Comment (benwillig) [inline mis_builder_budget/models/mis_report_instance_period.py:50]: I think we should have a single method which get the analytic filters from the context, here we have the logic in ´``mis_builder/models/mis_report_instance``` and ´``mis_builder_budget/models/mis_report_instance_period```. What do you think @sbidoul?

Comment (sbidoul) [reply mis_builder_budget/models/mis_report_instance_period.py:50]: @benwillig indeed, it would be better. Can you do it?

Comment (benwillig) [reply mis_builder_budget/models/mis_report_instance_period.py:50]: Done

**Comment by sbidoul** [comment]
@jalzaga thanks a lot for your valuable tests! Your last issue should now be resolved. Can you test again?

**Comment by fclementic2c** [comment]
functional test ok :+1:

**Review by fclementic2c — APPROVED** [review]

**Comment by sbidoul** [comment]
@benwillig can you update your review, it's still in change requested state?

**Review by jalzaga — APPROVED** [review]
Checked. Result  OK. The analytic filter is restablished to null when the field is blanked out.

**Comment by HaraldPanten** [comment]
In my honest (and functional) opinion, this PR is ready to be merged.

Thx a lot for your great deal of effort!

**Review by benwillig — APPROVED** [review]

**Comment by sbidoul** [comment]
:fireworks: : Thanks to all who contributed in a way or another!

I'll try to forward port to 11 over the week-end.

**Comment by sbidoul** [comment]
But if anyone is willing to port the changes to the widget (@mpanarin? @benwillig?) he'll earn my eternal gratitude :)

**Comment by fclementic2c** [comment]
@sbidoul I might need you eternal gratitude...especially if eternal....  :)
Yep we take this one.

**Comment by jalzaga** [comment]
Now the next step is adding the analytic tags filtering.

**Comment by sbidoul** [comment]
@jalzaga for analytic tags please create first an issue explaining what exactly you want to achieve.

**Comment by fclementic2c** [comment]
Filter on tags would be interesting feature if you want to use tags as analytic dimensions as I do.
but this filter has to be a "multi-select" since it is a M2M field...

**Comment by sbidoul** [comment]
Please open an issue for analytic tags. Also explaining what the criteria is when the user selects more than one tag.

**Pull request merged**